src/llvm-late-gc-lowering.cpp

```diff
@@ -1889,6 +1902,189 @@ void LateLowerGCFrame::PlaceGCFrameStores(State &S, unsigned MinColorRoot,
+    bool all_edge_in_set = true;
+    for (auto pred_bb: predecessors(edge.second)) {
+        if (edge.first == pred_bb)
+            continue;
+        if (edges.count(std::make_pair(pred_bb, edge.second)) == 0) {
+            all_edge_in_set = false;
+            break;
+        }
+    }
+    if (all_edge_in_set) {
+        for (auto pred_bb: predecessors(edge.second)) {
+            if (edge.first == pred_bb)
+                continue;
+            edges.remove(std::make_pair(pred_bb, edge.second));
+        }
+        ins_vec.push_back(edge.second->getFirstNonPHI());
+        return;
+    }
+    all_edge_in_set = true;
+    for (auto succ_bb: successors(edge.first)) {
+        if (edge.second == succ_bb)
+            continue;
+        if (edges.count(std::make_pair(edge.first, succ_bb)) == 0) {
+            all_edge_in_set = false;
+            break;
+        }
+    }
+    if (all_edge_in_set) {
+        for (auto succ_bb: successors(edge.first)) {
+            if (edge.second == succ_bb)
+                continue;
+            edges.remove(std::make_pair(edge.first, succ_bb));
+        }
+        ins_vec.push_back(edge.first->getTerminator());
+        return;
+    }
+    ins_vec.push_back(SplitEdge(edge.first, edge.second, dt)->getTerminator());
+}
+
+std::pair<SmallVector<Instruction*,4>,SmallVector<Instruction*,16>>
+LateLowerGCFrame::LocatePushPop(State &S)
+{
+    std::pair<SmallVector<Instruction*,4>,SmallVector<Instruction*,16>> res;
+    auto &push = res.first;
+    auto &pop = res.second;
+    auto global_frame = [&] {
+        push.push_back(&*(++BasicBlock::iterator(ptlsStates)));
+        for (auto &BB: *S.F) {
+            if (auto ret = dyn_cast<ReturnInst>(BB.getTerminator())) {
+                pop.push_back(ret);
+            }
+        }
+    };
+    // These are harder to handle, ignore them for now...
+    if (!S.Allocas.empty() || !S.ReturnsTwice.empty()) {
+        global_frame();
+        return res;
+    }
+    // We don't want to place any push/pop in a loop so first identify the graph of SCC nodes
+    // Each SCC node will always has the same "liveness".
+    struct SCCInfo {
+        std::vector<BasicBlock*> BBs;
+        SetVector<unsigned> Succ{};
+        SetVector<unsigned> Pred{};
+        bool HasLive = false;
+        bool MayReachLive = false;
+        SCCInfo(const std::vector<BasicBlock*> &BBs)
+            : BBs(BBs)
+        {
+        }
+    };
+    SmallVector<SCCInfo,4> sccnodes;
+    std::map<BasicBlock*,unsigned> sccmap;
+    for (auto SCCI = scc_begin(S.F); !SCCI.isAtEnd(); ++SCCI) {
+        unsigned id = sccnodes.size();
+        sccnodes.emplace_back(*SCCI);
+        auto &node = sccnodes.back();
+        for (auto bb: node.BBs) {
+            sccmap[bb] = id;
+            BBState &BBS = S.BBStates[bb];
+            node.HasLive |= BBS.HasLive;
+            for (auto succ_bb: successors(bb)) {
+                auto succ_id = sccmap[succ_bb];
+                if (succ_id < id) {
```

Comment:
I think `succ_id` might be uninitialized here (e.g. 0). Perhaps offset `id` by one when you put it in the map and change the test here to `succ_id && succ_id != id`